ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+    // the number of bytes written out to the asyncWritableChannel. This would be the size of the blob eventually.
+    private Long bytesWritten = 0L;
+    // the number of chunks that have been written out to the asyncWritableChannel.
+    private final AtomicInteger numChunksWrittenOut = new AtomicInteger(0);
+    // the index of the next chunk that is to be written out to the asyncWritableChannel.
+    private int indexOfNextChunkToWriteOut = 0;
+    // whether this object has called the readIntoCallback yet.
+    private AtomicBoolean readIntoCallbackCalled = new AtomicBoolean(false);
+    // the callback that is passed into the asyncWritableChannel write() operation.
+    Callback<Long> chunkAsyncWriteCallback = new Callback<Long>() {
+      @Override
+      public void onCompletion(Long result, Exception exception) {
+        bytesWritten += result;
+        if (exception != null) {
+          operationException.set(exception);
+        }
+        numChunksWrittenOut.incrementAndGet();
+      }
+    };
+
+    /**
+     * The bytes that will be read from this channel is not known until the read is complete.
+     * @return size written when called after the readInto is complete, -1 any time before that.
+     */
+    @Override
+    public long getSize() {
+      return streamSize;
+    }
+
+    @Override
+    public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
+      if (readCalled) {
+        throw new IllegalStateException("Cannot read the result of a GetBlob operation more than once");
+      }
+      this.asyncWritableChannel = asyncWritableChannel;
+      readIntoCallback = callback;
+      readIntoFuture = new FutureResult<>();
+      readCalled = true;
+      if (operationException.get() != null) {
+        completeRead();
+      }
+      return readIntoFuture;
+    }
+
+    @Override
+    public boolean isOpen() {
+      return isOpen;
+    }
+
+    @Override
+    public void close()
+        throws IOException {
+      isOpen = false;
+    }
+
+    /**
+     * @return whether readInto() has been called yet.
+     */
+    boolean isReadCalled() {
+      return readCalled;
+    }
+
+    /**
+     * @return the number of chunks that have been written out to the {@link AsyncWritableChannel}
+     */
+    int getNumChunksWrittenOut() {
+      return numChunksWrittenOut.get();
+    }
+
+    /**
+     * Attempt to write the data associated with the blob to the channel passed in by the caller (if the caller has
+     * done so).
+     */
+    private void maybeWriteToChannel() {
+      // if there are chunks available to be written out, do now.
+      if (firstChunk.isComplete() && readCalled) {
+        while (operationException.get() == null && chunkIndexToBuffer.containsKey(indexOfNextChunkToWriteOut)) {
+          ByteBuffer chunkBuf = chunkIndexToBuffer.remove(indexOfNextChunkToWriteOut);
+          asyncWritableChannel.write(chunkBuf, chunkAsyncWriteCallback);
```

Comment:
If that happens, it will bring down the router today. Although `RunTimeException` should not ideally happen unless there is a programming error, we can discuss on what the right behavior should be. In the worst case, we could have a catch all within the operation managers that simply removes the bad operation from its list.